### PR TITLE
SDF-modulated per-octave vol RFF PE scaling (sdf_pe_mlp)

### PR DESCRIPTION
**Pull request by morganmcg1**
## Hypothesis

The Random Fourier Feature (RFF) positional encoding used for volume points treats all spatial frequencies uniformly, regardless of a point's distance from the vehicle surface. Near-surface volume points are geometry-critical — they set pressure gradients that propagate outward — but far-field points have flatter gradients where high-frequency RFF components carry less signal. We hypothesize that **modulating per-octave RFF feature magnitudes by SDF distance** will let the model learn to attenuate high-frequency encoding noise far from the surface, improving vol_p generalization.

The SDF value is already available as `vol_x[:, :, 3]` (the 4th coordinate channel of raw volume inputs, after x/y/z). The model currently uses `pos-encoding-mode=string_separable` with `rff-num-features=16` and `rff-init-sigmas="0.25,0.5,1.0,2.0,4.0"` (5 octaves × 3 axes = 15 feature groups, each with 16 features = 240 features total, projected to hidden_dim=512).

The change is: after computing the 5-octave RFF features, apply a learned SDF-conditioned per-octave scalar weight `w_k = sdf_pe_mlp(sdf)[k]` to each of the 5 octave blocks before concatenation. The MLP is tiny (~112 params): `Linear(1→16) → SiLU → Linear(16→5) → Softplus`. Softplus guarantees positivity (we are scaling, not gating). Initialize `Linear(16→5)` to output all-ones so the baseline PE is preserved at step 0.

**Expected mechanism:** Low SDF (near surface) → model learns to upweight fine-grained octaves; high SDF (far field) → model downweights high-frequency noise. The OOD test_vol_p gap (~3× val/test multiplier) may partly arise from the model over-relying on high-frequency positional cues that are geometry-specific to training cars. Learned SDF-modulated attenuation could generalize better to unseen geometries.

**Parameter cost:** ~112 params. Purely a model.py change (~20 lines). No change to training script CLI.

## Instructions

Add a new CLI flag `--use-sdf-pe-scaling` (default: False) to `train.py` and implement the feature in `model.py`.

### Implementation in `model.py`

1. Add an `SdfPeScaler` module (or inline in the volume encoder):

```python
class SdfPeScaler(nn.Module):
    """Learns per-SDF-distance per-octave RFF scaling weights."""
    def __init__(self, num_octaves: int = 5):
        super().__init__()
        self.mlp = nn.Sequential(
            nn.Linear(1, 16),
            nn.SiLU(),
            nn.Linear(16, num_octaves),
            nn.Softplus(),
        )
        # Initialize final layer so Softplus(0) ≈ ln(2) ≈ 0.693
        # Use a small negative bias so output ≈ 1.0 at init
        # Softplus(x) → 1.0 when x ≈ 0.541; init bias to 0.541
        nn.init.zeros_(self.mlp[2].weight)
        nn.init.constant_(self.mlp[2].bias, 0.541)

    def forward(self, sdf: torch.Tensor) -> torch.Tensor:
        # sdf: [B, N, 1]
        return self.mlp(sdf)  # [B, N, num_octaves]
```

2. In the volume encoder's `forward`, after computing the per-octave RFF blocks and before concatenating them:

```python
# rff_features is a list of 5 tensors, each [B, N, 2*rff_num_features]
# sdf is vol_x[:, :, 3:4]  (raw SDF channel)
if self.use_sdf_pe_scaling:
    scales = self.sdf_pe_scaler(sdf)  # [B, N, 5]
    rff_features = [
        feat * scales[:, :, k:k+1]
        for k, feat in enumerate(rff_features)
    ]
```

3. Wire `use_sdf_pe_scaling=True/False` through the model constructor and the train.py argument parser.

### Run configuration

Run a 4-epoch screen (single arm, no ablation needed — if NEGATIVE we move on):

```bash
cd target/ && torchrun --standalone --nproc_per_node=8 train.py \
  --agent askeladd --optimizer lion --lr 9e-5 --weight-decay 5e-4 \
  --tau-y-loss-weight 1.5 --tau-z-loss-weight 2.0 --surface-loss-weight 2.0 \
  --ema-decay 0.999 --grad-clip-norm 0.5 --lr-warmup-epochs 1 \
  --pos-encoding-mode string_separable --use-qk-norm \
  --rff-num-features 16 --rff-init-sigmas "0.25,0.5,1.0,2.0,4.0" \
  --lr-cosine-t-max 4 --epochs 4 \
  --vol-points-schedule "0:16384:1:32768:2:49152:3:65536" \
  --no-compile-model \
  --model-layers 5 --model-hidden-dim 512 --model-heads 4 --model-slices 128 \
  --batch-size 4 --validation-every 1 \
  --train-surface-points 65536 --eval-surface-points 65536 \
  --train-volume-points 65536 --eval-volume-points 65536 \
  --use-surf-to-vol-xattn \
  --use-sdf-pe-scaling \
  --wandb-group askeladd-sdf-modulated-vol-pe
```

### EP gates (4-ep screen, steps align at 10,864 / 16,300 / 19,926 / 22,644)

- **EP1** (step ≈ 10,864): kill if val_abupt > 35%
- **EP2** (step ≈ 16,300): kill if val_abupt > 11.5% (baseline EP2 ≈ 7.938%)
- **EP3** (step ≈ 19,926): kill if val_abupt > 8.0% AND vol_p > 5.5% (both must fail; either passing means continue)
- **EP4** terminal: report val_abupt, vol_p, surface_pressure, wall_shear

If EP3 passes and EP4 val_abupt < 6.2868% (single-model SOTA), notify advisor for 13-epoch full run.

## Baseline

**Single-model SOTA (target to beat):** val_abupt=6.2868%, test_abupt=7.7445%
- volume_pressure: val=3.9152%, test=12.0063%
- surface_pressure: val=4.1766%, test=3.9100%
- wall_shear: val=7.0476%, test=6.9848%
- **W&B run:** `29nohj67` (group `nezuko-vol-aux-decoder-head`)
- **PR:** #958

**SOTA reproduce (for reference — your run adds `--use-sdf-pe-scaling`):**
```bash
cd target/ && torchrun --standalone --nproc_per_node=8 train.py \
  --agent nezuko --optimizer lion --lr 9e-5 --weight-decay 5e-4 \
  --tau-y-loss-weight 1.5 --tau-z-loss-weight 2.0 --surface-loss-weight 2.0 \
  --ema-decay 0.999 --grad-clip-norm 0.5 --lr-warmup-epochs 1 \
  --pos-encoding-mode string_separable --use-qk-norm \
  --rff-num-features 16 --rff-init-sigmas "0.25,0.5,1.0,2.0,4.0" \
  --lr-cosine-t-max 13 --epochs 13 \
  --vol-points-schedule "0:16384:3:32768:6:49152:9:65536" \
  --no-compile-model \
  --model-layers 5 --model-hidden-dim 512 --model-heads 4 --model-slices 128 \
  --batch-size 4 --validation-every 1 \
  --train-surface-points 65536 --eval-surface-points 65536 \
  --train-volume-points 65536 --eval-volume-points 65536 \
  --use-surf-to-vol-xattn \
  --use-vol-pressure-aux-head --volume-loss-weight 1.0 \
  --wandb-group nezuko-vol-aux-decoder-head
```

## Terminal result format

Post exactly one line (outside any code block) when done:

```
SENPAI-RESULT: {"terminal":true,"status":"complete","pending_arms":false,"wandb_run_ids":["<run-id>"],"primary_metric":{"name":"val_primary/abupt_axis_mean_rel_l2_pct","value":<ep4_value>},"test_metric":{"name":"test_primary/abupt_axis_mean_rel_l2_pct","value":<test_value_or_null>}}
```

### CODE REVIEW

**Comment by morganmcg1** [comment]
ADVISOR: Closing PR #1004 because ADVISOR: Closing this PR — it duplicates the already fully-closed SDF-modulated vol PE axis (PR #989, run z6dcbe9g, EP4 val_abupt=7.490%). PR #989 demonstrated that the SDF-PE MLP learns physically coherent near-surface amplification but its far-field attenuation (mean scale=0.157 at sdf=+2.0m) degrades far-field vol point discrimination, ending 1.0–1.1pp worse than the single-model SOTA. This axis is exhausted. Reassigning askeladd to a genuinely open hypothesis..
